Nrfx 6553 fix tls handling for s2ram

adamkondraciuk opened the pull request:
The S2RAM procedure requires marker checking after reset. Such checking is performed on very early stage of the system initialization and must ensure that the stack is not used due to the TLS pointer which is not initialized yet.

adamkondraciuk committed:
arch: arm: cortex_m: pm_s2ram: Rework S2RAM mark functions

The S2RAM procedure requires marker checking after reset.
Such checking is performed on very early stage of the system initialization
and must ensure that the stack is not used due to the TLS pointer which is
not initialized yet.

Signed-off-by: Adam Kondraciuk <adam.kondraciuk@nordicsemi.no>

nordic-krch commented on arch/arm/core/cortex_m/pm_s2ram.S:
```diff
@@ -67,10 +114,16 @@ SECTION_FUNC(TEXT, arch_pm_s2ram_suspend)
+	 * Save the return address from `pm_s2ram_mark_set`
+	 */
+	mov r1, r15
+	add r1, #6
```
Can you add comment why 6 is added? is this 3 16 bit instructions (mov add b)? `B` can only jump +-2kB. Maybe it would be better to store LR in R1 and use branch with link?

adamkondraciuk replied:
> Maybe it would be better to store LR in R1 and use branch with link?
Sounds reasonable - done

nordic-krch commented on include/zephyr/arch/common/pm_s2ram.h:
```diff
@@ -76,6 +81,11 @@ void pm_s2ram_mark_set(void);
  * allows custom implementation.
+ * The following requirements must be fulfilled:
+ * - the function cannot use stack (most likely asm function)
+ * - the content of the R1, R4 and LR registers must remain unchanged
```
why R4?

adamkondraciuk replied:
> why R4?

The content of R0 was moved there:
https://github.com/zephyrproject-rtos/zephyr/pull/80039/files#diff-011b9cab66d0e92083260695202e01cc69700c2b3283e29ae196735bfc39c753L31-L32
And restored after calling the function:
https://github.com/zephyrproject-rtos/zephyr/pull/80039/files#diff-011b9cab66d0e92083260695202e01cc69700c2b3283e29ae196735bfc39c753R151

nordic-krch commented on soc/nordic/common/pm_s2ram.S:
```diff
@@ -0,0 +1,73 @@
+	/*
+	 * Load check RESETREAS register
+	 */
+	ldr	r2, =CONFIG_NRF_RESETINFO_PERIPH_ADDRESS
```
Ideally we should avoid putting MDK stuff in kconfig. What about using inline assembly in c file? Then you can provide data from MDK.

adamkondraciuk replied:
> Ideally we should avoid putting MDK stuff in kconfig. What about using inline assembly in c file? Then you can provide data from MDK.

Can we be sure that such a function won't push anything onto the stack?

nordic-krch reviewed (commented):
Please remove `[nrf noup]` from commit messages.

adamkondraciuk committed:
soc: nordic: s2ram: Align s2ram marking procedures

Rework Nordic specific S2RAM marking procedures.
The S2RAM marking procedures must not disrupt the stack due to
the TLS pointer not yet being initialized during their execution.

Signed-off-by: Adam Kondraciuk <adam.kondraciuk@nordicsemi.no>

ithinuel reviewed (approved):
It is a bit unfortunate that we need to resort to ASM but I understand why.

If anything, a minor nit would be to use a register > r3 so save LR (eg `r5`) just to stay more "inline" with the usual ABI (which has R0-R3 caller saved).

mathieuchopstm commented on include/zephyr/arch/common/pm_s2ram.h:
```diff
@@ -65,6 +65,11 @@ int arch_pm_s2ram_suspend(pm_s2ram_system_off_fn_t system_off);
  *
  * Default implementation is setting a magic word in RAM. CONFIG_PM_S2RAM_CUSTOM_MARKING
  * allows custom implementation.
+ * The following requirements must be fulfilled:
+ * - the function cannot use stack (asm function or function with 'naked' attribute)
+ * - the content of the R1 and R4 registers must remain unchanged
+ * - returning from the function should be performed with the `bx lr` instruction
```
This feels out of place: this header is supposed to be arch-agnostic, but these restrictions are ARM-specific.

Also, as much as I understand the restrictions on `pm_s2ram_mark_check_and_clear`, I don't see why they should apply `pm_s2ram_mark_set` (since it runs on the stack of a regular thread, post OS boot, etc).

stephanosio replied:
exactly my thought. this is simply wrong.

stephanosio reviewed (commented):
I know I am a bit late to the party; but, I believe this should be reverted.

First of all, ARM arch-specific details must not be added to an `arch/common` header, as already noted above.

Second, the changes here seem to completely disregard the AAPCS conventions not taking into account caller saved and callee saved registers, also why are you taking a parameter via `r4`?

There seems to exist [a special and completely arbitrary contract between Cortex-M `arch_pm_s2ram_suspend` and `pm_s2ram_mark_set`](https://github.com/zephyrproject-rtos/zephyr/pull/80039#discussion_r1808462895) -- this is just madness.

TBH, the entire Cortex-M S2RAM implementation seems to be a big mess and should have never been accepted upstream in that condition ...
